feat: add cosmere initiative system

jsc17 opened the pull request:
New PR due to changing to the 0.1.0 branch:

Overrides Foundry's default initiative rolling system to use the Cosmere RPGs slow/fast popcorn style initiative system. Implements a bucket/phases system to group combatants based on turn speed and whether they are a player character or an adversary. Defaults all combatants to slow turn speed when they are added to the combat. Added buttons to the combat tracker allowing them to switch turn speed and indicate they have acted for the turn, and set some basic styling for the combat-tracker template.

Done:

* Create flags on combatant creation to store turn speed and activation status, defaulting to slow for the turn speed.
* Set initiative based on turn speed/character type, instead of rolling
* Add buttons to combat tracker to switch turn speed and indicate the character has activated for the turn (using default foundry icons as placeholders)
* Add headers to the combat tracker to sort player/npc turns (Fast Players, Fast NPCs, etc ...)
* Add context menu option to combat tracker combatants to reset activation status
* Set localization for all text values

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,151 @@
+
+export default class CosmereCombatTracker extends CombatTracker {
+    //overrides default tracker template to implement slow/fast buckets and activation button.
+    //eslint-disable-next-line -- eslint rules want readonly field
```
When using `eslint-disable-next-line` you can specify which rule to disable. While it may not have any further effect here, we should make sure to only disable specific rules when its needed.

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,151 @@
+import CosmereCombatant from './combatant';
+
+export default class CosmereCombatTracker extends CombatTracker {
+    //overrides default tracker template to implement slow/fast buckets and activation button.
```
This might be nice to include as a doc comment on the whole class

jsc17 replied:
Moved the comment to the whole class, please let me know if you want something more detailed. I'm used to working on solo hobby projects, so commenting is definitely an area I need to work on

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,151 @@
+            turns: CosmereTurn[];
+            [x: string]: unknown;
+        };
+        data.turns = data.turns.map((turn) => {
```
Is there a reason to type this as `[x: string]: unknown`? 
Couldn't we type the whole thing as:
```ts
turns: CosmereTurn[];
fastPlayers: CosmereTurn[];
slowPlayers: CosmereTurn[];
fastNPC: CosmereTurn[];
slowNPC: CosmereTurn[];
```

Keeping everything properly typed will make it easier for other people to make changes to this code should it be needed in the future.

jsc17 replied:
I had it unknown for something I was testing when trying to set it up, fixed

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,151 @@
+        );
+    }
+
+    //toggles combatant turn speed on clicking the "fast/slow" text on the combat tracker window
```
This can also be set up as a doc comment

jsc17 replied:
Not exactly sure what you mean here. Can you elaborate a bit?

stanavdb replied:
If you use JSDoc comments (comments starting with `/**`) instead of regular comments, then your comment will show up alongside intellisense on most IDEs. Have a look at the Actor document for some examples.

Broadly: Any comments that explain the functionality (for someone who might want to use that function/variable/parameter/etc.) should be doc comments. Any comments that explain the implementation should be regular comments.

jsc17 replied:
This was a pretty good explanation. I hadn't used JSDoc comments before so I didn't realize they did that for intellisense. I've expanded the comments in the code I've added.

stanavdb commented on src/system/applications/combat/combatant.ts:
```diff
@@ -0,0 +1,42 @@
+import { CosmereActor } from '../../documents';
+
+export default class CosmereCombatant extends Combatant {
+    //sets actor type to CosmereActor, to get typescript to stop yelling. Might be a better way to do this but I'm newish to TS and don't know how.
+    override get actor(): CosmereActor {
+        return super.actor as CosmereActor;
+    }
```
This is a perfectly fine way to re-declare the type

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,151 @@
+    pending: number;
+    finished: number;
+    type?: string;
+    turnSpeed?: string;
```
Turn speed should be limited to only be `'fast'` or `'slow'`. You can do this by either making it a literal type or by using an enum.

**Literal**
```ts
type TurnSpeed = 'fast' | 'slow';
```

**Enum**
```ts
const enum TurnSpeed {
    Fast = 'fast',
    Slow = 'slow'
}
```

stanavdb commented on src/system/applications/combat/combatant.ts:
```diff
@@ -0,0 +1,42 @@
+        );
+    }
+
+    generateInitiative(type: string, speed: string): number {
```
See above about turn speed typing. Actor type enum is available from `@system/types/cosmere`.

jsc17 replied:
fixed as suggested

stanavdb reviewed (changes requested):
No real comments on the implementation. Some minor style/typescript comments & tips. However when I run it locally I get an error when adding a combatant to a combat. 

`Error: An error occurred while rendering CosmereCombatTracker 24. The partial combatant could not be found`

It looks like the combatant partial is not getting preloaded. (Maybe something when wrong when merging with release-0.1.0?) Preloading of partials should be set up in the `preloadHandlebarsTemplates` function, found at: `util/handlebars/index.ts`

jsc17 commented:
Yeah, when switching to the 0.1.0 branch I moved the template into preloadHandleBarTemplates, but didn't commit the updated function. Fixed that and implemented most of the other changes you mentioned.

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,156 @@
+            newTurn.css = '';
+            return newTurn;
+        });
+        data.fastPlayers = data.turns.filter((turn) => {
+            return turn.type == 'character' && turn.turnSpeed == 'fast';
+        });
+        data.slowPlayers = data.turns.filter((turn) => {
+            return turn.type == 'character' && turn.turnSpeed == 'slow';
+        });
+        data.fastNPC = data.turns.filter((turn) => {
+            return turn.type == 'adversary' && turn.turnSpeed == 'fast';
+        });
+        data.slowNPC = data.turns.filter((turn) => {
+            return turn.type == 'adversary' && turn.turnSpeed == 'slow';
+        });
```
These should use the `ActorType` and `TurnSpeed` enums for the comparison. Also its generally best practice to use `===` instead of `==`.

jsc17 replied:
Fixed enums and replaced all ==

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,156 @@
+    css: string;
+    pending: number;
+    finished: number;
+    type?: string;
+    turnSpeed?: string;
```
Turn speed and type are still typed as string here

stanavdb commented on src/system/applications/combat/combat.ts:
I don't know how I didn't realise this the first time around, but both `CosmereCombat` and `CosmereCombatant` are Documents in Foundry and should be under the system/documents, not system/applications

jsc17 replied:
I was following along with a different modules structure and didn't really know the difference. From what I understand now, documents are data (like actors) and applications will have ui components (like the character sheets). I know there's more to it than that, but is that essentially correct?

stanavdb commented on src/system/applications/combat/combatant.ts:
See comment combat

stanavdb commented on src/system/applications/combat/combatant.ts:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    generateInitiative(type: ActorType, speed: TurnSpeed): number {
+        let initiative: number = this.actor.system.attributes.spd.value;
```
No need to type initiative as `number` here. `actor.system.attributes.spd.value` is already typed as a number so it should infer the type automatically

stanavdb commented on src/system/applications/combat/combatant.ts:
```diff
@@ -0,0 +1,54 @@
+            this.id!,
+            this.generateInitiative(
+                this.actor.type as ActorType,
+                TurnSpeed.Slow,
```
Looking at this I realise I hadn't re-declared the type for actors to ActorType yet. If you add the following to the top of the CosmereActor class you can get rid of all the `as ActorType` casts. 

```ts
    // Redeclare `actor.type` to specifically be of `ActorType`.
    // This way we avoid casting everytime we want to check/use its type
    declare type: ActorType;
```

See system/documents/item for reference.

stanavdb commented on src/templates/combat/combatant.hbs:
```diff
@@ -0,0 +1,50 @@
+    <div class="token-name flexcol">
+        <h4>{{this.name}}</h4>
+        <div class="combatant-controls flexrow">
+            {{#if ../user.isGM}}
+            <a class="combatant-control {{#if this.hidden}}active{{/if}}" aria-label="{{localize 'COMBAT.ToggleVis'}}"
+                role="button" data-tooltip="COMBAT.ToggleVis" data-control="toggleHidden">
+                <i class="fas fa-eye-slash"></i>
+            </a>
+            <a class="combatant-control {{#if this.defeated}}active{{/if}}"
+                aria-label="{{localize 'COMBAT.ToggleDead'}}" role="button" data-tooltip="COMBAT.ToggleDead"
+                data-control="toggleDefeated">
+                <i class="fas fa-skull"></i>
+            </a>
+            {{/if}}
```
This doesn't appear to work for me. I'm connected as a GM user but I don't see the visibility toggle. Maybe you need to `@root.user.isGM`? Or perhaps the context isn't getting pass through properly because it's a partial.

jsc17 replied:
This appears to have been something between them being in an each block and a partial. Adding @root to the appropriate sections resolved the issue.

stanavdb commented on src/templates/combat/combatant.hbs:
```diff
@@ -0,0 +1,50 @@
+        <span class="resource">{{this.resource}}</span>
+    </div>
+    {{/if}}
+    <div class="cosmere-combatant-buttons">
+        <a class="cosmere-turn-speed-control"></a>
+            {{#if this.activated}}
+            <a class="cosmere-activate-control-activated"></a>
+            {{else}}
+            <a class="cosmere-activate-control"></a>
+            {{/if}}
+    </div>
```
Having tried it, it would be nice to have some tooltips (using `data-tooltip` attr) that indicate what the buttons do

stanavdb reviewed (changes requested):
Noticed a few more things that I'd missed/hadn't been able to check yet last time.

stanavdb commented on src/system/applications/combat/combat-tracker.ts:
```diff
@@ -0,0 +1,156 @@
+ * Overrides default tracker template to implement slow/fast buckets and activation button.
+ */
+export default class CosmereCombatTracker extends CombatTracker {
+    // eslint-disable-next-line @typescript-eslint/class-literal-property-style
```
As a good habit, when I disable an eslint rule I add a note comment (`// NOTE: ...`) describing why that rule had to be disabled. This is useful for when other people look at the code (or yourself after a long while away) and forces you to really justify needing to disable eslint rules. 
Disabling eslint rules should be generally discouraged, but is sometimes needed
